runtime/sema/check_array_expression.go

```diff
@@ -70,22 +74,26 @@ func (checker *Checker) VisitArrayExpression(expression *ast.ArrayExpression) as
 
 		checker.checkVariableMove(value)
 		checker.checkResourceMoveOperation(value, valueType)
-
-		// infer element type from first element
-		// TODO: find common super type?
-		if elementType == nil {
-			elementType = valueType
-		}
 	}
 
 	checker.Elaboration.ArrayExpressionArgumentTypes[expression] = argumentTypes
 
 	if elementType == nil {
-		// i.e: contextually expected type is not available and array has zero elements.
-		elementType = NeverType
-	}
+		// Contextually expected type is not available.
+		// Therefore, find the least common supertype of the elements.
+		elementType = LeastCommonSuperType(argumentTypes...)
+
+		if elementType == InvalidType {
+			checker.report(
+				&TypeAnnotationRequiredError{
+					Cause: "cannot infer type from array literal: ",
```

Comment:
```suggestion
					Cause: "cannot infer type from array literal",
```

Or was the idea to add something after the colon?

Reply:
yes, this is to get a more readable error message, since `TypeAnnotationRequiredError` appends some more texts to the cause.

https://github.com/onflow/cadence/blob/81c63bfef73cd4874d05b6274dbd2402a0a0b788/runtime/sema/errors.go#L2883-L2888